lib/OnyxUtils.ts

```diff
@@ -397,6 +427,24 @@ function addToEvictionBlockList(key: OnyxKey, connectionID: number): void {
 
     evictionBlocklist[key].push(connectionID);
 }
+/**
+ * It extracts the non-numeric collection identifier of a given key.
+ *
+ * For example:
+ * - `getCollectionKey("report_123")` would return "report_"
+ * - `getCollectionKey("report")` would return "report"
+ * - `getCollectionKey("report_")` would return "report_"
+ * - `getCollectionKey(null)` would return ""
+ *
+ * @param {OnyxKey} key - The key to process.
+ * @return {string} The pure key without any numeric
+ */
+function getCollectionKey(key: OnyxKey): string {
+    if (!key) {
+        return '';
+    }
+    return key.replace(/_\w+/g, '_');
+}
```

Comment:
@chrispader @gedu Soooo as part of the workspace feeds and billing docs (I think) we are actually going to add onyx collection keys that have 2 `_` for example `sharedNVP_private_expensifyCardSettings_12345`

Will these changes work with it? cc @fabioh8010

Reply:
No this will not work in this case. We would need to search for the last `_` and only remove the rest after that. cc @gedu

Reply:
Raised [here](https://expensify.slack.com/archives/C05LX9D6E07/p1719310826344159)

Reply:
```suggestion
function getCollectionKey(key: OnyxKey): string {
    if (!key) {
        return '';
    }
    const lastUnderscoreIndex = key.lastIndexOf('_');
    if (lastUnderscoreIndex === -1) {
        return key;
    }
    return key.substring(0, lastUnderscoreIndex);
}
```

Reply:
won't work, will be always the last one?

Reply:
Maybe we could use fix and use `splitCollectionMemberKey` instead? Regardless of the decision, could we use this PR to fix `splitCollectionMemberKey` to use `lastIndexOf()`?

Reply:
> won't work, will be always the last one?

Good question. I think we should have some convention around this. Allowing underscores in both the collection key and the member key might cause problems where the correct key cannot be determined correctly.

Imagine having a collection `some_collection_` and the member key `member_key1`. Which would result in the full key `some_collection_member_key1`.

In this case it would not work...

Reply:
I think we can have underscore in non-collection keys like some private_nvpname. We cannot easily change that since that is how its stored in the db.

But we can have the convention that the collection has _<number> in the end I think

Reply:
@mountiny what should I do? I got lost

Reply:
I updated and added some tests

Reply:
> Instead we should scrap the "collection part" based on the valid/allowed collection keys configured here:

I think this might not be that easy either as we are going to add more complex collection keys

I think for now that a reasonable way to detect a collection key is that after the last `_`, there is only one number left. So that would be a way to detect a collection key

If you have a key that has _ in it and its not a collection, it would not have any number at the end

Reply:
Or we can add sharedNVP_...... to the collection keys

Reply:
@mountiny  I think in this PR we should keep just this collection update, and in a follow up PR look into the new single key with `_`, I'm seeing that we will need to plan that part a bit more